### PR TITLE
Added cityscapes dataset

### LINKED ISSUE #214: [data request] <Cityscapes>
* Name of dataset: Cityscapes
* URL of dataset: https://www.cityscapes-dataset.com/
* License of dataset: This dataset is made freely available to academic and non-academic entities for non-commercial purposes such as academic research, teaching, scientific publications, or personal experimentation. Permission is granted to use the data given that you agree:

That the dataset comes “AS IS”, without express or implied warranty. Although every effort has been made to ensure accuracy, we (Daimler AG, MPI Informatics, TU Darmstadt) do not accept any responsibility for errors or omissions.
That you include a reference to the Cityscapes Dataset in any work that makes use of the dataset. For research papers, cite our preferred publication as listed on our website; for other media cite our preferred publication as listed on our website or link to the Cityscapes website.
That you do not distribute this dataset or modified versions. It is permissible to distribute derivative works in as far as they are abstract representations of this dataset (such as models trained on it or additional annotations that do not directly include any of our data) and do not allow to recover the dataset or something similar in character.
That you may not use the dataset or any derivative work for commercial purposes as, for example, licensing or selling the data, or using the data with a purpose to procure a commercial gain.
That all rights not expressly granted to you are reserved by us (Daimler AG, MPI Informatics, TU Darmstadt).
* Short description of dataset and use case(s): We present a new large-scale dataset that contains a diverse set of stereo video sequences recorded in street scenes from 50 different cities, with high quality pixel-level annotations of 5 000 frames in addition to a larger set of 20 000 weakly annotated frames. The dataset is thus an order of magnitude larger than similar previous attempts. Details on annotated classes and examples of our annotations are available at this webpage.

The Cityscapes Dataset is intended for

1. assessing the performance of vision algorithms for two major tasks of semantic urban scene understanding: pixel-level and instance-level semantic labeling;

2. supporting research that aims to exploit large volumes of (weakly) annotated data, e.g. for training deep neural networks.

Folks who would also like to see this dataset in `tensorflow/datasets`, please thumbs-up so the developers can know which requests to prioritize.

And if you'd like to contribute the dataset (thank you!), see our [guide to adding a dataset](https://github.com/tensorflow/datasets/blob/master/docs/add_dataset.md).

### PR DESCRIPTION
**Pull request by f1recracker**
Added Cityscapes dataset with segmentation labels. Fixes #214.

### CODE REVIEW

**Comment by f1recracker** [comment]
Hello! This is my first contribution to tensorflow-datasets! I actually had a few quick questions / design items to discuss before merging this.

1. How are the unit tests run? (is it via `oss_scripts/oss_tests`?) I haven't run any of the cityscapes test but I checked the size of using the actual dataset in my local repository.

2. What is the recommended size of fake datasets used for testing? I have tried to ensure each split has a different size but I still end up with a 38MB fake dataset (cityscapes has big images - 1024 * 2048).

**Comment by f1recracker** [comment]
I've used pngquant to compress the train images!

**Comment by us** [comment]
@f1recracker some dataset's test generate own images and label like `testing/cifar.py`. Maybe it can reduce size even more.

**Comment by cyfra** [comment]
@f1recracker - for running the tests, you can also invoke them from command line directly (assuming you're in the top directory):

python3 -m tensorflow_datasets.image.celeba_test

File: tensorflow_datasets/image/cityscapes.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+  def _split_generators(self, dl_manager):
+    paths = {
+        'images': os.path.join(dl_manager.manual_dir, 'leftImg8bit_trainvaltest.zip'),
+        'labels': os.path.join(dl_manager.manual_dir, 'gtFine_trainvaltest.zip'),
+    }
+
+    if any(not os.path.exists(p) for p in paths.values()):
+      msg = 'You must download the dataset files manually and place them in: '
+      msg += ', '.join(paths.values())
```
Comment (cyfra) [inline tensorflow_datasets/image/cityscapes.py]: can you add some explanation on why the tfds cannot download the files?

Comment (f1recracker) [reply tensorflow_datasets/image/cityscapes.py]: Cityscapes (much like ImageNet) requires a login before you can download the files.

Comment (cyfra) [reply tensorflow_datasets/image/cityscapes.py]: :-(  I see. please mention it in the comment here.

File: tensorflow_datasets/image/cityscapes.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
+# pylint: disable=bad-indentation
```
Comment (cyfra) [inline tensorflow_datasets/image/cityscapes.py]: any reason why ? ;-)

File: tensorflow_datasets/image/cityscapes.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+  consisting of coarse labels for training and validation splits as well as 19998 images
+  in a new 'train_extra' split which may be useful for data-heavy applications.
+'''
+
+# TODO add instance ids (might need to import cityScapesScripts)
+
+class CityscapesCore(tfds.core.GeneratorBasedBuilder):
+  '''Base for cityscapes datasets'''
+
+  VERSION = tfds.core.Version('2.0.0')
```
Comment (cyfra) [inline tensorflow_datasets/image/cityscapes.py]: the 2.0 - is it on purpose? (by default, we start with 1.0 versions for the new dataset)

File: tensorflow_datasets/image/cityscapes.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        features=tfds.features.FeaturesDict({
+            'image': tfds.features.Image(shape=(1024, 2048, 3), encoding_format='png'),
+            'label': tfds.features.Image(shape=(1024, 2048, 1), encoding_format='png'),
+            'image_id': tfds.features.Text(),
+        }),
+        supervised_keys=('image', 'label'),
+        urls=["https://www.cityscapes-dataset.com", "https://github.com/mcordts/cityscapesScripts"],
+        citation=_CITATION,
+    )
+
```
Comment (cyfra) [inline tensorflow_datasets/image/cityscapes.py]: Add @abc.abstractmethod

File: tensorflow_datasets/image/cityscapes.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+  author={Cordts, Marius and Omran, Mohamed and Ramos, Sebastian and Rehfeld, Timo and Enzweiler, Markus and Benenson, Rodrigo and Franke, Uwe and Roth, Stefan and Schiele, Bernt},
+  booktitle={Proc. of the IEEE Conference on Computer Vision and Pattern Recognition (CVPR)},
+  year={2016}
+}
+'''
+
+_DESCRIPTION = '''\
+  Cityscapes is a dataset consisting of diverse urban street scenes and dense pixel-level
+  annotations taken across 50 different cities at varying times of the year. The label
+  annotations span across 30 (34 including ingore indices) distinct classes of objects
```
Comment (cyfra) [inline tensorflow_datasets/image/cityscapes.py]: typo 'ingore -> ignored' ?

File: tensorflow_datasets/image/cityscapes.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+  commonly encountered in driving scenes. Detailed label information may be found here:
+  https://github.com/mcordts/cityscapesScripts/blob/master/cityscapesscripts/helpers/labels.py#L52-L99
+
+  The dataset consists of 5000 annotated image, label pairs at 1024 * 2048 resolution
+  pre-split into training (2975) validation (500) and test (1525) sets.
```
Comment (cyfra) [inline tensorflow_datasets/image/cityscapes.py]: Please add:

WARNING: this dataset requires users to setup a login and password in order to get the files.

File: tensorflow_datasets/image/cityscapes.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+            },
+        ),
+    ]
+
+# Helper functions
+
+IMAGE_FILE_RE = re.compile(r'([a-z\-]+)_(\d+)_(\d+)_leftImg8bit\.png')
+
+def _image_id(image_file):
+  match = IMAGE_FILE_RE.match(image_file)
```
Comment (cyfra) [inline tensorflow_datasets/image/cityscapes.py]: add comment on what this function does.

File: tensorflow_datasets/image/cityscapes.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+    if any(not os.path.exists(p) for p in paths.values()):
+      msg = 'You must download the dataset files manually and place them in: '
+      msg += ', '.join(paths.values())
+      raise AssertionError(msg)
+
+    paths['images'] = os.path.join(dl_manager.extract(paths['images']), 'leftImg8bit')
+    paths['labels'] = os.path.join(dl_manager.extract(paths['labels']), 'gtFine')
+
+    # Number of shards calculations:
+    # Max shard size = 4GB
```
Comment (cyfra) [inline tensorflow_datasets/image/cityscapes.py]: very nice - thanks.

File: tensorflow_datasets/image/cityscapes_test.py [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+class CityscapesTest(testing.DatasetBuilderTestCase):
+  DATASET_CLASS = cityscapes.Cityscapes
+  SPLITS = {  # Expected number of examples on each split.
+      'train': 3,
+      'validation': 1,
+      'test': 2,
+  }
+
+  DATASET_CLASS = cityscapes.CityscapesCoarse
```
Comment (cyfra) [inline tensorflow_datasets/image/cityscapes_test.py]: does it work? I think that you might need to create another class in this file (otherwise you might be just overwriting the entry from line 12)

Comment (f1recracker) [reply tensorflow_datasets/image/cityscapes_test.py]: Yes this is incorrect! I wasn't sure how to run the unit tests until today and I need another class.

File: tensorflow_datasets/image/cityscapes.py (hunk 2 above)
Comment (f1recracker) [reply tensorflow_datasets/image/cityscapes.py]: Sorry! My global pylint config uses 4 spaces which causes a lot of false positives in Google style python code! Will remove!

File: tensorflow_datasets/image/cityscapes.py (hunk 3 above)
Comment (f1recracker) [reply tensorflow_datasets/image/cityscapes.py]: No I just used the latest I could find from other examples. Will move to 1.0

File: tensorflow_datasets/image/cityscapes.py (hunk 6 above)
Comment (f1recracker) [reply tensorflow_datasets/image/cityscapes.py]: Where would this go? end of the first paragraph? end of description string?

Comment (cyfra) [reply tensorflow_datasets/image/cityscapes.py]: Wherever you prefer. 
You could add it at the end.

Later it will be automatically copied to datasets.md

**Comment by f1recracker** [comment]
I've fixed most of the changes but I have observed that `dl_manager.extract` does not work as intended in unit tests (which actually returns DL_EXTRACT_RESULT or the example_path from what I can see here ) but `dl_manager.iter_archive` works. Is this intentional? I'm actually rewriting my code using `dl_manager.iter_archive`.

**Comment by cyfra** [comment]
@f1recracker - yes, you're right, this is an inconsistency in our testing framework that we introduced when adding DL_EXTRACT_RESULT.


@pierrot0 - Pierre, can you comment on the extract issue and what is the recommended way @f1recracker should follow?

**Comment by f1recracker** [comment]
Has there been any updates on this? Sorry I got a little caught up with some work during the past few weeks!

**Comment by f1recracker** [comment]
It seems that the latest version fixed the inconsistency and also another error. I've also refactored my code to now use BuilderConfigs instead of having separate classes for Cityscapes - fine and Cityscapes - coarse.

**Comment by lgeiger** [comment]
@f1recracker Thanks for the PR! Would it be also possible to add the depth data of the Cityscapes dataset as an additional feature?

**Comment by f1recracker** [comment]
I wasn't aware of the depth maps - let me have a look at this.

**Comment by lgeiger** [comment]
Thanks for taking a look!

**Comment by lgeiger** [comment]
@f1recracker The stereo pair right images would also be a great addition to this PR.

**Comment by f1recracker** [comment]
@lgeiger Added support for right images and disparity maps too. Cityscapes now has 4 configurations:
- 'semantic_segmentation' - left + segmentations
- 'semantic_segmentation_extra' - left + segmentations + coarse labels + train extra split
- 'stereo_disparity' - left + right + disparity maps
- 'stereo_disparity_extra' - left + right + disparity maps + train extra split

Could this be reviewed again? There's significant changes since the last review (BuilderConfigs so no abcs, and the recent change with disparity maps)? Thanks!

File: tensorflow_datasets/image/cityscapes.py [hunk 10]
```diff
@@ -194,24 +207,24 @@ def calculate_num_shards(split_size, features_size_mb):
...
-          num_shards=calculate_num_shards(500, features_size_mb),
-          gen_kwargs={
-              feat:os.path.join(path, 'val')
-              for feat, path in paths.items()
-              if not feat.endswith('/extra')
-          },
-      ),
+        tfds.core.SplitGenerator(
+            name=tfds.Split.TRAIN,
+            num_shards=calculate_num_shards(2975, features_size_mb),
```
Comment (us) [inline tensorflow_datasets/image/cityscapes.py]: You don't need to specify num_shards anymore.

Comment (f1recracker) [reply tensorflow_datasets/image/cityscapes.py]: Wasn't aware of this - thanks!

**Comment by netw0rkf10w** [comment]
@f1recracker Thanks for working on this.
@cyfra What would need to be changed for this to be merged please?

**Comment by cyfra** [comment]
Hey @netw0rkf10w 
Unfortunately we have grown quite a backlog of PRs - sorry for that. 
We were quite busy with making sure that current ones will work with Python3.

as I see that there is high interest with this dataset, I'll try to merge it soon.

**Comment by netw0rkf10w** [comment]
@cyfra Thank you for your reply, and for your hard work! This is indeed a very popular dataset in computer vision, so the merge would benefit a lot of users. Looking forward to having this merged.

File: tensorflow_datasets/image/cityscapes.py [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+        }
+
+        if self.builder_config.right_images:
+          features['image_right'] = os.path.join(
+              paths_city_root['images_right'], f'{image_id}_rightImg8bit.png')
+
+        if self.builder_config.segmentation_labels:
+          features['segmentation_label'] = os.path.join(
+              paths_city_root['segmentation_labels'],
+              f'{image_id}_{self.builder_config.label_suffix}.png')
```
Comment (Conchylicultor) [inline tensorflow_datasets/image/cityscapes.py]: This is using python3 only feature, while TFDS should be both Py2 and 3 compatible.

Comment (hermannsblum) [reply tensorflow_datasets/image/cityscapes.py]: @f1recracker https://github.com/f1recracker/tensorflow-datasets/pull/1 here is a possible fix

File: tensorflow_datasets/image/cityscapes.py [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+          ('disparity_trainextra.zip', 'disparity')
+        self.ignored_ids.add('troisdorf_000000_000073') # No disparity for this file
+
+
+class Cityscapes(tfds.core.GeneratorBasedBuilder):
+  '''Base class for Cityscapes datasets'''
+
+  BUILDER_CONFIGS = [
+      CityscapesConfig(
+          name='semantic_segmentation',
```
[the remark this answers is not in the document]
Comment (f1recracker) [reply tensorflow_datasets/image/cityscapes.py]: It was a deliberate choice - Disparity maps need additional permissions from the authors and aren't readily available. 'Extra' splits - whether for segmentation or for the disparity requires additional files (50GB). I chose this route so that each configuration so that it would work for all cases.

File: tensorflow_datasets/image/cityscapes.py (hunk 11 above)
Comment (f1recracker) [reply tensorflow_datasets/image/cityscapes.py]: will fix tonight. Thanks

File: tensorflow_datasets/image/cityscapes.py [hunk 13]
```diff
@@ -0,0 +1,280 @@
...
+      ),
+  ]
+
+  VERSION = tfds.core.Version('1.0.0')
+
+  def _info(self):
+    # Enable features as necessary
+    features = {}
+    features['image_id'] = tfds.features.Text()
+    features['image_left'] =\
```
[the remark this answers is not in the document]
Comment (f1recracker) [reply tensorflow_datasets/image/cityscapes.py]: will fix tonight. Thanks.

File: tensorflow_datasets/image/cityscapes.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (f1recracker) [reply tensorflow_datasets/image/cityscapes.py]: I gave this some more thought today to see if I could refactor the code to use fewer if/else statements. The problem arises because Cityscapes has too many features that every user may not use.

1. There's two 'datasets' for different tasks - semantic semgentation and disparity inference. Disparity tasks can also require the right images - another additional download.
2. Every cityscapes 'dataset' (semantic seg / stereo depth) has an additional 'extra' split. This split is much larger and any labels provided are coarse grain instead.

I think we still need 4 (2 (task) x 2 (extra split)) configurations.

One feature that might reduce some if/else might be to use a constant feature dict and use a 'default' image tensor (since cityscape labels are images) if this is acceptable.

**Comment by netw0rkf10w** [comment]
Thanks a lot, @f1recracker, for continuing working on this! Unfortunately I am not yet familiar enough with the code to be able to help you.

**Comment by cyfra** [comment]
Ok - dataset is submitted. 

Thanks a lot @f1recracker for your contribution and patience.

File: tensorflow_datasets/image/cityscapes.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Aya-S) [reply tensorflow_datasets/image/cityscapes.py]: Are there any tutorials on how to use this class please?

[the remark this answers is not in the document]
Comment (f1recracker) [reply tensorflow_datasets/image/cityscapes.py]: You just have to do something along the lines of:

```
import tensorflow_datasets as tfds
dataset = tfds.load(name="cityscapes/semantic_segmentation_extra", split="train")
# Can also specify config from above.
```

I'd give a working Colab but unfortunately this is a pretty big dataset and requires manual download :)

[the remark this answers is not in the document]
Comment (Aya-S) [reply tensorflow_datasets/image/cityscapes.py]: @f1recracker thank you for the reply! 
I downloaded the dataset manually and will attempt to access it with the syntax you suggested
Thanks again.
